Micronaut won't start (throws exception) when "discovery-client" feature is enabled

### Expected Behavior

The application should start.

### Actual Behaviour

The _BeanInstantiationException_ exception is thrown.

Stack trace:
```
19:20:06.684 [main] INFO  i.m.context.DefaultBeanContext - Reading bootstrap environment configuration
19:20:06.848 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [io.micronaut.discovery.eureka.EurekaServiceInstanceList] could
not be loaded: io/micronaut/http/client/HttpClientConfiguration
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [io.micronaut.discovery.eureka.EurekaServiceInstanceList] could not be loaded: io/micronaut/http/client/HttpClientConfiguration
        at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.getDefinition(DefaultBeanContext.java:4157)
        at io.micronaut.context.DefaultBeanContext.collectBeanCandidates(DefaultBeanContext.java:2138)
        at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:2113)
        at io.micronaut.context.DefaultBeanContext.findBeanCandidatesInternal(DefaultBeanContext.java:3362)
        at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3417)
        at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1436)
        at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:881)
        at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:871)
        at io.micronaut.context.event.ApplicationEventPublisherFactory$2.lambda$$1(ApplicationEventPublisherFactory.java:215)
        at io.micronaut.core.util.SupplierUtil$1.get(SupplierUtil.java:47)
        at io.micronaut.context.event.ApplicationEventPublisherFactory$2.publishEvent(ApplicationEventPublisherFactory.java:226)
        at io.micronaut.context.DefaultBeanContext.publishEvent(DefaultBeanContext.java:1779)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:360)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:198)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:73)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:322)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:308)
        at foo.Application.main(Application.java:8)
Caused by: java.lang.NoClassDefFoundError: io/micronaut/http/client/HttpClientConfiguration
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
        at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
        at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
        at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
        at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        at io.micronaut.discovery.eureka.$EurekaServiceInstanceList$Definition.<clinit>(Unknown Source)
        at io.micronaut.discovery.eureka.$EurekaServiceInstanceList$Definition$Reference.load(Unknown Source)
        at io.micronaut.context.AbstractInitializableBeanDefinitionReference.load(AbstractInitializableBeanDefinitionReference.java:178)
        at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.getDefinition(DefaultBeanContext.java:4152)
        ... 17 common frames omitted
Caused by: java.lang.ClassNotFoundException: io.micronaut.http.client.HttpClientConfiguration
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        ... 39 common frames omitted
```


### Steps To Reproduce

Create a new micronaut application with _discovery-client_ feature.
Run the application.

```
mn create-app foo.bar --features=discovery-client --build=maven --lang=java

cd bar

mvn mn:run
```

### Environment Information

OS: Windows 11 / WSL 2 (Ubuntu)

```
$ java -version
openjdk version "17.0.8.1" 2023-08-24
OpenJDK Runtime Environment Temurin-17.0.8.1+1 (build 17.0.8.1+1)
OpenJDK 64-Bit Server VM Temurin-17.0.8.1+1 (build 17.0.8.1+1, mixed mode, sharing)

mvn --version
Apache Maven 3.9.4 (dfbb324ad4a7c8fb0bf182e6d91b0ae20e3d2dd9)
Maven home: /home/user/.sdkman/candidates/maven/current
Java version: 17.0.8.1, vendor: Eclipse Adoptium, runtime: /home/user/.sdkman/candidates/java/17.0.8.1-tem
Default locale: en, platform encoding: UTF-8
OS name: "linux", version: "5.15.90.1-microsoft-standard-wsl2", arch: "amd64", family: "unix"

mn --version
Micronaut Version: 4.1.1
```

### Example Application

_No response_

### Version

4.1.1

--- a/starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java
+++ b/starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java
@@ -23,11 +23,7 @@ import io.micronaut.starter.application.OperatingSystem;
 import io.micronaut.starter.application.Project;
 import io.micronaut.starter.build.BuildPlugin;
 import io.micronaut.starter.build.BuildProperties;
-import io.micronaut.starter.build.dependencies.Coordinate;
-import io.micronaut.starter.build.dependencies.CoordinateResolver;
-import io.micronaut.starter.build.dependencies.Dependency;
-import io.micronaut.starter.build.dependencies.DependencyContext;
-import io.micronaut.starter.build.dependencies.LookupFailedException;
+import io.micronaut.starter.build.dependencies.*;
 import io.micronaut.starter.feature.Feature;
 import io.micronaut.starter.feature.Features;
 import io.micronaut.starter.feature.build.maven.Profile;
@@ -418,5 +414,14 @@ public class GeneratorContext implements DependencyContext {
 
     public boolean isMicronautFramework() {
         return getFramework().equalsIgnoreCase(Options.FRAMEWORK_MICRONAUT);
+    }
+
+    public boolean hasDependencyInScope(@NonNull String groupId,
+                                        @NonNull String artifactId,
+                                        @NonNull Scope scope) {
+        return getDependencies().stream()
+                .anyMatch(dependency -> dependency.getGroupId().equals(groupId) &&
+                        dependency.getArtifactId().equals(artifactId)
+                        && dependency.getScope() == scope);
     }
 }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryClient.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryClient.java
@@ -22,20 +22,40 @@ import io.micronaut.starter.build.dependencies.Dependency;
 import io.micronaut.starter.build.dependencies.MicronautDependencyUtils;
 import io.micronaut.starter.feature.Category;
 import io.micronaut.starter.feature.Feature;
+import io.micronaut.starter.feature.FeatureContext;
+import io.micronaut.starter.feature.httpclient.HttpClientFeature;
+import io.micronaut.starter.feature.other.HttpClient;
+import jakarta.inject.Inject;
 import jakarta.inject.Singleton;
 
 @Singleton
 public class DiscoveryClient implements Feature {
+    public static final String NAME = "discovery-client";
     private static final String ARTIFACT_ID_MICRONAUT_DISCOVERY_CLIENT = "micronaut-discovery-client";
     private static final Dependency DEPENDENCY_MICRONAUT_DISCOVERY_CLIENT = MicronautDependencyUtils.discovery()
             .artifactId(ARTIFACT_ID_MICRONAUT_DISCOVERY_CLIENT)
             .compile()
             .build();
 
+    private final HttpClient httpClient;
+
+    @Inject
+    public DiscoveryClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * @deprecated Use {@link DiscoveryClient(HttpClient)} instead.
+     */
+    @Deprecated
+    public DiscoveryClient() {
+        this(new HttpClient());
+    }
+
     @NonNull
     @Override
     public String getName() {
-        return "discovery-client";
+        return NAME;
     }
 
     @Override
@@ -46,6 +66,13 @@ public class DiscoveryClient implements Feature {
     @Override
     public String getDescription() {
         return "Adds micronaut-discovery-client dependency which provides an implmentation of the DiscoveryClient API.";
+    }
+
+    @Override
+    public void processSelectedFeatures(FeatureContext featureContext) {
+        if (!featureContext.isPresent(HttpClientFeature.class)) {
+            featureContext.addFeature(httpClient);
+        }
     }
 
     public void apply(GeneratorContext generatorContext) {

--- a/starter-core/src/main/java/io/micronaut/starter/feature/httpclient/HttpClientJdk.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/httpclient/HttpClientJdk.java
@@ -28,8 +28,9 @@ import java.util.List;
 public class HttpClientJdk implements HttpClientFeature {
     public static final String NAME = "http-client-jdk";
 
+    public static final String ARTIFACT_ID_MICRONAUT_HTTP_CLIENT_JDK = "micronaut-http-client-jdk";
     private static final Dependency DEPENDENCY_MICRONAUT_HTTP_CLIENT_JDK = MicronautDependencyUtils.coreDependency()
-            .artifactId("micronaut-http-client-jdk")
+            .artifactId(ARTIFACT_ID_MICRONAUT_HTTP_CLIENT_JDK)
             .compile()
             .build();
 

--- a/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
@@ -15,16 +15,19 @@
  */
 package io.micronaut.starter.feature.other;
 
+import io.micronaut.core.annotation.NonNull;
 import io.micronaut.starter.application.ApplicationType;
 import io.micronaut.starter.application.generator.GeneratorContext;
 import io.micronaut.starter.build.dependencies.Dependency;
 import io.micronaut.starter.build.dependencies.MicronautDependencyUtils;
+import io.micronaut.starter.build.dependencies.Scope;
 import io.micronaut.starter.feature.DefaultFeature;
 import io.micronaut.starter.feature.Feature;
 import io.micronaut.starter.feature.awslambdacustomruntime.AwsLambdaCustomRuntime;
 import io.micronaut.starter.feature.function.awslambda.AwsLambda;
 import io.micronaut.starter.feature.graalvm.GraalVM;
 import io.micronaut.starter.feature.httpclient.HttpClientFeature;
+import io.micronaut.starter.feature.httpclient.HttpClientJdk;
 import io.micronaut.starter.feature.validator.MicronautHttpValidation;
 import io.micronaut.starter.options.Options;
 import jakarta.inject.Singleton;
@@ -67,13 +70,20 @@ public class HttpClientTest implements DefaultFeature {
 
     @Override
     public void apply(GeneratorContext generatorContext) {
-        if (generatorContext.getFeatures().hasFeature(AwsLambdaCustomRuntime.class) || (generatorContext.getFeatures().hasFeature(AwsLambda.class) && generatorContext.getFeatures().hasFeature(GraalVM.class))) {
-            generatorContext.addDependency(HttpClient.DEPENDENCY_MICRONAUT_HTTP_CLIENT);
-        } else if (generatorContext.getApplicationType() == ApplicationType.DEFAULT) {
-            generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_TEST);
-            if (generatorContext.hasFeature(MicronautHttpValidation.class) && generatorContext.getBuildTool().isGradle()) {
-                generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_COMPILE_ONLY);
+        if (!hasHttpClientFeatureDependencyInScope(generatorContext, Scope.COMPILE)) {
+            if (generatorContext.getFeatures().hasFeature(AwsLambdaCustomRuntime.class) || (generatorContext.getFeatures().hasFeature(AwsLambda.class) && generatorContext.getFeatures().hasFeature(GraalVM.class))) {
+                generatorContext.addDependency(HttpClient.DEPENDENCY_MICRONAUT_HTTP_CLIENT);
+            } else if (generatorContext.getApplicationType() == ApplicationType.DEFAULT) {
+                generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_TEST);
+                if (generatorContext.hasFeature(MicronautHttpValidation.class) && generatorContext.getBuildTool().isGradle()) {
+                    generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_COMPILE_ONLY);
+                }
             }
         }
+    }
+
+    private boolean hasHttpClientFeatureDependencyInScope(@NonNull GeneratorContext generatorContext, @NonNull Scope scope) {
+        return generatorContext.hasDependencyInScope(MicronautDependencyUtils.GROUP_ID_MICRONAUT, ARTIFACT_ID_MICRONAUT_HTTP_CLIENT, scope) ||
+                generatorContext.hasDependencyInScope(MicronautDependencyUtils.GROUP_ID_MICRONAUT, HttpClientJdk.ARTIFACT_ID_MICRONAUT_HTTP_CLIENT_JDK, scope);
     }
 }

--- a/starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryClientSpec.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryClientSpec.groovy
@@ -1,0 +1,44 @@
+package io.micronaut.starter.feature.discovery
+
+import io.micronaut.starter.BeanContextSpec
+import io.micronaut.starter.BuildBuilder
+import io.micronaut.starter.build.BuildTestUtil
+import io.micronaut.starter.build.BuildTestVerifier
+import io.micronaut.starter.build.dependencies.Scope
+import io.micronaut.starter.feature.httpclient.HttpClientJdk
+import io.micronaut.starter.options.BuildTool
+
+class DiscoveryClientSpec extends BeanContextSpec {
+    void "dependencies for discovery client feature"(BuildTool buildTool) {
+        when:
+        String template = new BuildBuilder(beanContext, buildTool)
+                .features([DiscoveryClient.NAME])
+                .render()
+        BuildTestVerifier verifier = BuildTestUtil.verifier(buildTool, template)
+
+        then:
+        verifier.hasDependency("io.micronaut.discovery", "micronaut-discovery-client", Scope.COMPILE)
+        verifier.hasDependency("io.micronaut", "micronaut-http-client", Scope.COMPILE)
+        !verifier.hasDependency("io.micronaut", "micronaut-http-client", Scope.TEST)
+
+        where:
+        buildTool << BuildTool.values()
+    }
+
+    void "dependencies for discovery client and http-client-jdk"(BuildTool buildTool) {
+        when:
+        String template = new BuildBuilder(beanContext, buildTool)
+                .features([DiscoveryClient.NAME, HttpClientJdk.NAME])
+                .render()
+        BuildTestVerifier verifier = BuildTestUtil.verifier(buildTool, template)
+
+        then:
+        verifier.hasDependency("io.micronaut.discovery", "micronaut-discovery-client", Scope.COMPILE)
+        verifier.hasDependency("io.micronaut", HttpClientJdk.ARTIFACT_ID_MICRONAUT_HTTP_CLIENT_JDK, Scope.COMPILE)
+        !verifier.hasDependency("io.micronaut", "micronaut-http-client", Scope.COMPILE)
+        !verifier.hasDependency("io.micronaut", "micronaut-http-client", Scope.TEST)
+
+        where:
+        buildTool << BuildTool.values()
+    }
+}

--- a/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
+++ b/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
@@ -4,7 +4,6 @@ import io.micronaut.starter.options.BuildTool
 import io.micronaut.starter.options.Language
 import io.micronaut.starter.test.CommandSpec
 import io.micronaut.starter.test.LanguageBuildCombinations
-import spock.lang.Unroll
 
 class DiscoveryClientFunctionalSpec extends CommandSpec {
 

--- a/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
+++ b/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
@@ -7,7 +7,6 @@ import io.micronaut.starter.test.LanguageBuildCombinations
 
 class DiscoveryClientFunctionalSpec extends CommandSpec {
 
-    @Unroll
     void 'test discovery-client for #lang and #buildTool'(Language lang, BuildTool buildTool) {
         given:
         generateProject(lang, buildTool, ['discovery-client'])

--- a/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
+++ b/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/discovery/DiscoveryClientFunctionalSpec.groovy
@@ -1,0 +1,30 @@
+package io.micronaut.starter.core.test.feature.discovery
+
+import io.micronaut.starter.options.BuildTool
+import io.micronaut.starter.options.Language
+import io.micronaut.starter.test.CommandSpec
+import io.micronaut.starter.test.LanguageBuildCombinations
+import spock.lang.Unroll
+
+class DiscoveryClientFunctionalSpec extends CommandSpec {
+
+    @Unroll
+    void 'test discovery-client for #lang and #buildTool'(Language lang, BuildTool buildTool) {
+        given:
+        generateProject(lang, buildTool, ['discovery-client'])
+
+        when:
+        String output = executeBuild(buildTool, "test")
+
+        then:
+        output.contains("BUILD SUCCESS")
+
+        where:
+        [lang, buildTool] << LanguageBuildCombinations.combinations()
+    }
+
+    @Override
+    String getTempDirectoryPrefix() {
+        return "test-app-discovery-client"
+    }
+}

